servers/src/main/java/tachyon/worker/file/FileDataManager.java

```diff
@@ -75,36 +78,49 @@ public synchronized void persistFile(long fileId, List<Long> blockIds) throws IO
     OutputStream outputStream = mUfs.create(dstPath);
     final WritableByteChannel outputChannel = Channels.newChannel(outputStream);
 
-    for (long blockId : blockIds) {
-      long lockId;
-      try {
-        lockId = mBlockDataManager.lockBlock(Sessions.CHECKPOINT_SESSION_ID, blockId);
-      } catch (BlockDoesNotExistException e) {
-        throw new IOException(e);
+    Map<Long, Long> blockIdToLockId = Maps.newHashMap();
+    List<Throwable> errors = new ArrayList<Throwable>();
+    try {
+      // lock all the blocks to prevent any eviction
+      for (long blockId : blockIds) {
+        long lockId = mBlockDataManager.lockBlock(Sessions.CHECKPOINT_SESSION_ID, blockId);
+        blockIdToLockId.put(blockId, lockId);
       }
 
-      // obtain block reader
-      try {
+      for (long blockId : blockIds) {
+        long lockId = blockIdToLockId.get(blockId);
+
+        // obtain block reader
         BlockReader reader;
-        try {
-          reader =
-              mBlockDataManager.readBlockRemote(Sessions.CHECKPOINT_SESSION_ID, blockId, lockId);
-        } catch (BlockDoesNotExistException e) {
-          throw new IOException(e);
-        } catch (InvalidWorkerStateException e) {
-          throw new IOException(e);
-        }
+        reader = mBlockDataManager.readBlockRemote(Sessions.CHECKPOINT_SESSION_ID, blockId, lockId);
 
         // write content out
         ReadableByteChannel inputChannel = reader.getChannel();
         BufferUtils.fastCopy(inputChannel, outputChannel);
         reader.close();
-      } finally {
+      }
+    } catch (BlockDoesNotExistException e) {
+      errors.add(e);
+    } catch (InvalidWorkerStateException e) {
+      errors.add(e);
+    } finally {
+      // make sure all the locks are released
+      for (long lockId : blockIdToLockId.values()) {
         try {
           mBlockDataManager.unlockBlock(lockId);
-        } catch (BlockDoesNotExistException e) {
-          throw new IOException(e);
+        } catch (BlockDoesNotExistException bdnee) {
+          errors.add(bdnee);
+        }
+      }
+
+      // there're errors on releasing locks
+      if (!errors.isEmpty()) {
+        StringBuilder errorStr = new StringBuilder();
+        errorStr.append("the blocks of file").append(fileId).append(" are failed persist\n");
+        for (Throwable e : errors) {
+          errorStr.append(e).append('\n');
```

Comment:
It's possible to have more than one exceptions. For example, it can buffer multiple errors thrown from unlocking blocks.